### PR TITLE
core: checked arithmetics

### LINKED ISSUE #2555: Audit for all usage of unchecked arithmetic and update nomenclature
Looking through the codebase for unchecked arithmetic I still find quite a bit, e.g.:
- [here](https://github.com/anoma/namada/blob/4209a7413c21df3f9b1190189990f4146c00dda3/crates/core/src/types/dec.rs#L368)
- [here](https://github.com/anoma/namada/blob/4209a7413c21df3f9b1190189990f4146c00dda3/crates/core/src/types/uint.rs#L667)
- many more, just searching...

We should:
- Remove ALL support for unchecked arithmetic. We should never use it.
- Update all arithmetic to use `checked_sub`, `checked_add` etc. names (don't overload `+` / `-`) to make it clear that we're using checked arithmetic (and so a search for only `+` or `-`, for example, should return only internal implementation code instead decimal types and no usage other than that).

### PR DESCRIPTION
**Pull request by tzemanovic**
## Describe your changes

first part of #2555 addressing core crate

## Indicate on which release or other PRs this topic is based on

v0.34.0

## Checklist before merging to `draft`
- [x] I have added a changelog
- [x] Git history is in acceptable state

### CODE REVIEW

**Comment by cwgoes** [comment]
@tzemanovic Is the title of this PR still accurate? Looks like a bunch of crates beyond core are changed. Does this cover everything, everywhere or are there other sections still remaining to transition to checked arithemetic?

File: crates/core/src/dec.rs [hunk 1]
```diff
@@ -285,17 +335,18 @@ impl TryFrom<String> for Dec {
...
-            None => Self::zero(),
-        }
+impl TryFrom<token::Amount> for Dec {
+    type Error = Error;
+
+    fn try_from(amt: token::Amount) -> std::result::Result<Self, Self::Error> {
+        let raw = I256::try_from(amt.raw_amount())
+            .map_err(|e| eyre!("Invalid raw amount: {e}"))?;
+        let denom = I256(Uint::exp10(
+            (POS_DECIMAL_PRECISION - NATIVE_MAX_DECIMAL_PLACES) as usize,
```
Comment (brentstone) [inline crates/core/src/dec.rs:345]: I'm sure we're safe to do this operation but figured I'd point it out as possibly unchecked

File: crates/core/src/token.rs [hunk 2]
```diff
@@ -670,17 +734,17 @@ impl From<Amount> for U256 {
     }
 }
 
-impl From<Dec> for Amount {
-    fn from(dec: Dec) -> Amount {
+impl TryFrom<Dec> for Amount {
+    type Error = ();
```
Comment (brentstone) [inline crates/core/src/token.rs:738]: why `()` for the `Error` type?

File: crates/core/src/token.rs [hunk 3]
```diff
@@ -870,7 +800,7 @@ pub enum AmountParseError {
         "Error decoding token amount, too many decimal places: {0}. Maximum \
          {1}"
     )]
-    ScaleTooLarge(u32, u8),
+    ScaleTooLarge(usize, u8),
```
Comment (brentstone) [inline crates/core/src/token.rs:803]: should we avoid this in light of #826?

File: crates/core/src/token.rs [hunk 4]
```diff
@@ -1032,6 +965,78 @@ pub mod testing {
...
+        type Output = Self;
+
+        fn sub(self, rhs: Self) -> Self::Output {
+            self.checked_sub(rhs).unwrap()
+        }
+    }
+
+    impl std::ops::SubAssign for Amount {
+        fn sub_assign(&mut self, rhs: Self) {
+            *self = *self - rhs;
```
Comment (brentstone) [inline crates/core/src/token.rs:992]: why not checked?

File: crates/core/src/token.rs [hunk 5]
```diff
@@ -1032,6 +965,78 @@ pub mod testing {
...
+        fn mul(self, rhs: T) -> Self::Output {
+            self.checked_mul(rhs.into()).unwrap()
+        }
+    }
+
+    impl std::ops::Mul<Amount> for u64 {
+        type Output = Amount;
+
+        fn mul(self, rhs: Amount) -> Self::Output {
+            rhs * self
```
Comment (brentstone) [inline crates/core/src/token.rs:1011]: checked? And same for the ones below this

File: crates/proof_of_stake/src/epoched.rs [hunk 6]
```diff
@@ -106,7 +107,7 @@ where
             Some(last_update) => {
                 let data_handler = self.get_data_handler();
                 let future_most_epoch =
-                    last_update + FutureEpochs::value(params);
+                    last_update.unchecked_add(FutureEpochs::value(params));
```
Comment (brentstone) [inline crates/proof_of_stake/src/epoched.rs:110]: why unchecked add?

**Review by brentstone — CHANGES_REQUESTED** [review]
Just left some comments and questions, but otherwise LGTM

**Comment by tzemanovic** [comment]
> @tzemanovic Is the title of this PR still accurate? Looks like a bunch of crates beyond core are changed. Does this cover everything, everywhere or are there other sections still remaining to transition to checked arithemetic?

@cwgoes  It is accurate - the lints to warn about unchecked arithmetics and conversions are only applied in the core crate, i.e.:

```rust
#![warn(
    clippy::cast_sign_loss,
    clippy::cast_possible_truncation,
    clippy::cast_possible_wrap,
    clippy::cast_lossless,
    clippy::arithmetic_side_effects
)]
```

The changes outside of core are only due to removal of unchecked API in core types (except for a few reasonable exceptions), which are extensive but not exhaustive. We should also add these lints to other relevant crates which can still do unchecked arithmetics and conversions with other non-core types.

File: crates/core/src/dec.rs (hunk 1 above)
Comment (tzemanovic) [reply crates/core/src/dec.rs:345]: it's ok as they're both consts and the fn is also covered by unit tests

File: crates/core/src/token.rs (hunk 2 above)
Comment (tzemanovic) [reply crates/core/src/token.rs:738]: ah, I'll switch it to use arith err

File: crates/core/src/token.rs (hunk 3 above)
Comment (tzemanovic) [reply crates/core/src/token.rs:803]: it's ok in error type - it doesn't affect the state machine

File: crates/core/src/token.rs (hunk 4 above)
Comment (tzemanovic) [reply crates/core/src/token.rs:992]: these are all within the `mod testing` where it's ok and convenient

File: crates/core/src/token.rs (hunk 5 above)
Comment (tzemanovic) [reply crates/core/src/token.rs:1011]: as above

File: crates/core/src/uint.rs [hunk 7]
```diff
@@ -593,224 +597,212 @@ impl I256 {
...
+                Self(inner)
             }
             (false, false) => {
-                let this = self.abs();
-                let that = other.abs();
-                that.cmp(&this)
+                let inner = self.abs().checked_add(rhs.abs())?;
+                if inner > MAX_SIGNED_VALUE {
+                    return None;
+                }
```
[the remark this answers is not in the document]
Comment (tzemanovic) [reply crates/core/src/uint.rs:640]: I think this is already limited in the `MAX_SIGNED_VALUE` (`"Note the the last digit is 2^63 - 1"`)

File: crates/proof_of_stake/src/epoched.rs (hunk 6 above)
Comment (tzemanovic) [reply crates/proof_of_stake/src/epoched.rs:110]: This is for convenience when working with epochs - it's u64 so we should never overflow on additions with system params. I added this doc on the method: Panics on overflow. Care must be taken to only use this with trusted values that are known to be in a limited range (e.g. system parameters but not e.g. transaction variables).

File: crates/core/src/token.rs (hunk 2 above)
Comment (tzemanovic) [reply crates/core/src/token.rs]: fixed in [70daab5](https://github.com/anoma/namada/pull/3074/commits/70daab5da619e8028203c93d7197ef9d5970b184)

**Comment by cwgoes** [comment]
Understood @tzemanovic - thanks!

**Pull request merged**